ui/src/main/java/org/pentaho/di/ui/spoon/tree/provider/SlavesFolderProvider.java

```diff
@@ -52,23 +63,62 @@ public SlavesFolderProvider() {
+    try {
+      Set<String> projectSlaveServerNames = new HashSet<>();
+      // Bowl specific
+      if ( currentBowl != DefaultBowl.getInstance() ) {
+        //SlaveServerManagementInterface slaveServerManager = currentBowl.getManager( SlaveServerManagementInterface.class );
+        org.pentaho.di.shared.SharedObjectsManagementInterface slaveServerManager = currentBowl.getManager( org.pentaho.di.shared.SharedObjectsManagementInterface.class );
 
-    servers.sort( ( s1, s2 ) -> String.CASE_INSENSITIVE_ORDER.compare( s1.getName(), s2.getName() ) );
+        //List<SlaveServer> servers = slaveServerManager.getSlaveServers();
+        List<SharedObjectInterface> sharedObjects = slaveServerManager.getSharedObjects( SharedObjectsManager.SLAVESERVER_TYPE );
+        sharedObjects.sort( ( s1, s2 ) -> String.CASE_INSENSITIVE_ORDER.compare( s1.getName(), s2.getName() ) );
+        //for ( SlaveServer slaveServer : servers ) {
+        for ( SharedObjectInterface sharedObject : sharedObjects ) {
+          if ( !filterMatch( sharedObject.getName(), filter ) ) {
+            continue;
+          }
 
-    for ( SlaveServer slaveServer : servers ) {
-      if ( !filterMatch( slaveServer.getName(), filter ) ) {
-        continue;
+          projectSlaveServerNames.add( sharedObject.getName() );
+          TreeNode childTreeNode = createTreeNode( treeNode, sharedObject.getName(), guiResource.getImageSlaveTree(),
+                                  LeveledTreeNode.LEVEL.PROJECT, false );
+        }
       }
 
-      TreeNode childTreeNode = createTreeNode( treeNode, slaveServer.getName(), guiResource.getImageSlaveTree() );
-      if ( slaveServer.isShared() ) {
-        childTreeNode.setFont( guiResource.getFontBold() );
+      // Global
+      org.pentaho.di.shared.SharedObjectsManagementInterface globalServerMgr = DefaultBowl.getInstance().getManager( org.pentaho.di.shared.SharedObjectsManagementInterface.class );
+      Set<String> globalServerNames = new HashSet<>();
+      List<SharedObjectInterface> servers = globalServerMgr.getSharedObjects( SharedObjectsManager.SLAVESERVER_TYPE );
+      servers.sort( ( s1, s2 ) -> String.CASE_INSENSITIVE_ORDER.compare( s1.getName(), s2.getName() ) );
+      for ( SharedObjectInterface slaveServer : servers ) {
+        if ( !filterMatch( slaveServer.getName(), filter ) ) {
+          continue;
+        }
+
+        globalServerNames.add( slaveServer.getName() );
+        TreeNode childTreeNode = createTreeNode( treeNode, slaveServer.getName(), guiResource.getImageSlaveTree(),
+          LeveledTreeNode.LEVEL.GLOBAL, projectSlaveServerNames.contains( slaveServer.getName() ) );
       }
 
+      // Local SlaveServers
+      if ( meta.isPresent() ) {
+        SharedObjectsManagementInterface localServerMgr = meta.get().getSharedObjectsManagementInterface();
+        List<SharedObjectInterface> localServers = localServerMgr.getSharedObjects( SharedObjectsManager.SLAVESERVER_TYPE );
```

Comment:
This style, where the calling code has to pass a constant, rather than there being a specific manager instance for each type, will make it harder to refactor this to use the metastore later. I think we need a manager for each type. Those managers could be almost entirely implemented in a base class, but they have to each have a separate instance, I think, for the later conversion to the metastore to work the way we have discussed. 

Like, with the current DatabaseManagementInterface that we added, refactoring DatabaseManagementInterface to be implemented on top of a metastore is trivial, because the calling code doesn't know any of the internal implementation. This code will not be as easy because of the exposed internals.